File: Backtracking/39.Combination_Sum/C++/39.Combination_Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int> & candidates, int target) 
    {
        //==========================================//
        // Input type:                              //
        //   - vector<int> ref. (candidates vector) //
        //   - int (target value)                   //
        // Return type:                             //
        //   - vector<vector<int>> (result vector)  //
        //==========================================//

        /*Initialize*/
        ///// Length of candidates vector /////
        int len_candidates = candidates.size();

        ///// Record summary value /////
        int record_sum_val = 0;

        ///// Record vector /////
        vector<int> record_vect;

        ///// Result vector /////
        vector<vector<int>> res_vect;


        /*Recursion-based backtracking-based loop traversal*/
        sort(candidates.begin(), candidates.end()); //Sorted operations

        backtrackTraversal(len_candidates, candidates, target, 0, res_vect, record_vect); //Recursion function call

        return res_vect;
    }


    void backtrackTraversal(const int & len_candidates, const vector<int> & candidates, int target, int record_idx,
                            vector<vector<int>> & res_vect, vector<int> & record_vect)
    {
        //==================================================//
        // Input type:                                      //
        //   - const int ref. (length of candidates vector) //
        //   - const vector<int> ref. (candidates vector)   //
        //   - int (target value)                           //
        //   - int (record index)                           //
        //   - vector<vector<int>> ref. (result vector)     //
        //   - vector<int> ref. (record vector)             //
        // Return type:                                     //
        //   - void (no return)                             //
        //==================================================//
        if (record_idx >= len_candidates) { return; } //Issue, Boundary-case handling

        if (target < 0) { return; } //Issue, Boundary-case handling

        if (! target) { res_vect.push_back(record_vect); return; } //Issue, Boundary-case handling


        /*Whole process, flow*/
        for (int sub_idx = record_idx; (sub_idx < len_candidates); (sub_idx++)) //Whole
        {
            record_vect.emplace_back(candidates[sub_idx]); //Keep updating, recording

            backtrackTraversal(len_candidates, candidates, (target - candidates[sub_idx]), sub_idx, res_vect, record_vect); //Recursion function call

            record_vect.pop_back(); //Keep updating, popped

        } //Whole
    }
};

File: Backtracking/39.Combination_Sum/Python/39.Combination_Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        #============================================#
        # Recursion-based backtrack traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of candidates array #####
        len_candidates = len(candidates)

        ##### Result array #####
        res_arry = []


        ##########################################
        #Recursion-based Backtrack loop traversal
        candidates.sort() #Sorted operations

        self.backtrackTraversal(len_candidates, candidates, target, 0, [], res_arry) #Recursion function call

        return res_arry

        
    def backtrackTraversal(self, len_candidates, candidates, target, record_idx, record_arry, res_arry):
        """
        :type len_candidates: int
        :type candidates: List[int]
        :type target: int
        :type record_idx: int
        :type record_arry: List[int]
        :type res_arry: List[List[int]]
        :rtype: None, void
        """
        #==============================================#
        # Recursion-based backtrack traversal function #
        #==============================================#
        if (record_idx >= len_candidates): #Issue/Boundary-case handling
            return

        if (target < 0): #Issue/Boundary-case handling
            return

        if (not target): #Issue/Boundary-case handling
            res_arry.append(record_arry) #Update/Record

            return


        ####################
        #Whole process/flow
        for sub_idx in range(record_idx, len_candidates):
            self.backtrackTraversal(len_candidates, candidates, (target - candidates[sub_idx]), sub_idx, 
                                    (record_arry + [candidates[sub_idx]]), res_arry) #Recursion function call
